### ISSUE TITLE
Bot Adapters are not interchangingly usable

### ISSUE THREAD

**Issue by RamjotSingh** [Microsoft/botbuilder-dotnet#220]
**Problem** Right now Bot Adapters can't be interchangeably used due to ProcessActivity not being part of the base BotAdapter class. This means you can't just use a different BotAdapter while keeping the whole stack and unit test your code. This issue is more apparent in case of Dependency injection where ideally you should be able to just pass BotAdapter into controller and then for web you register BotFramework adapter in DI container and for unit tests just replace this adapter with TestAdapter.
This is because BotFramework adapter needs auth header to be passed in ProcessActivity.

**Solution, kinda?** - Make [IHttpContextAccessor](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.ihttpcontextaccessor?view=aspnetcore-2.0) a required param for BotFrameworkAdapter. This can be passed by just doing [new HttpContextAccessor()](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.httpcontextaccessor?view=aspnetcore-2.0) or registering it in DI container (it is registered by default in the ASPNETCore DI container I think). This way we can access the auth header directly in the adapter without the need for it to be passed explicitly.

*But*, this solution does not work on ASP.NET web apps (apps written using the web.config + Global.asax.cs way) because IHttpContextAccessor is a ASPNetCore concept and for legacy apps HttpContext is accessed using HttpContext.Current. This [HttpContext class](https://msdn.microsoft.com/en-us/library/system.web.httpcontext(v=vs.110).aspx) is defined in System.web.dll and is Windows specific. Even having different dll for both frameworks does not work either as a ASPNetCore project with target framework set to .NetFramework chooses the .NetFramework dll and not .NetStandard. So the only option is to have 2 packages altogether.

@tomlm @cleemullins

**Comment by drub0y** [comment]
**TL;DR; let the middleware be the awesomeness... everything up until the middleware is invoked should be host/integration specific**

No doubt, this is a leaky abstraction. I have different thoughts about how to address it, but, sticking to the specific topic of your solution, I believe that _any_ integration with ASP.NET should be done via the integration libraries (see #212 for WIP) which should introduce middleware to adapt the specific details of their host environments to the bot framework itself. 

When I initially stumbled across this early on I thought "why should `BotFrameworkAdapter` need an `authHeader`?" and then, when I saw what it was doing, "why in the heck is it doing anything with JWTs in there?". This doesn't seem like the right place for this responsibility to me. IMHO `BotContext` should expose a `Principal` property of type `ClaimsPrincipal`. Where those claims come from is something an upstream piece of authentication middleware should provide. Then downstream middleware and, ultimately, the bot itself can look into the claims to make other decisions about how the request can/should proceed.

As I mentioned earlier, in the case of ASP.NET (NetFx or Core) integration, I believe the integration libraries should come with their own baked in middleware to deal with this. It doesn't seem like a good idea to have the core framework knowing anything about authorization headers or processing JWTs. At most, the core framework should simply expect some claims to be found in the `ClaimsPrincipal` without any regard for how they got there.

Bigger picture, I actually think the adapter concept should be refactored into a "host" concept and possibly completely removed. It should likely be two distinct concepts:
 1.  Today it would remain to act as the runner of the middleware pipeline. However, when the great middleware refactoring happens you won't need a runner any more, you will literally invoke the outermost piece of middleware, so this concept should ultimately go away.
 2. Separate host integration libraries and each integration library has the following responsibilities:
   a. Receive an activity message through some transport
   b. Deserialize that message into an `Activity` instance
   c. Create the `BotContext` around that `Activity`
   d. Invoke the middleware with the component from step 1 passing through that `BotContext`

So `TestAdapter` should become `TestHost` and the `ConsoleAdapter` should become `ConsoleHost` and really there is nothing similar about how they perform steps a-c from above. Sure, they seem to do the same things so you might be tempted to find a common abstraction amongst them to put into a base `BotHost` class, but sometimes abstracting at the host level often boxes you into corners and bites you when it comes to integrating with a specific host that doesn't quite work the way your abstraction was designed. IMHO, let the host implementations all be their own unique snowflakes that best integrate those environments with the core bot framework.

**Comment by RamjotSingh** [comment]
**TL;DR; Instead of IBot let's just split the codebase into 3ish parts. One which is core platform agnostic, one which assumes HTTP pipeline, and Adapter.**

If this is the approach we are planning to take, doesn't that also come in sync with the whole multi-package thing (I actually thought we did not want to go down that route unless we had no other option left). 

I read thru the DCR (that was a good read:)) and ended up wonder why IBot? Why not just take the whole BotFrameworkAdapter, ConsoleAdapter, TestAdapter into 3 different packages and let people use these or write their own adapters? The pipeline sitting below (Middleware etc.) already are platform agnostic. This means that any dev can just take the Core package (the one with Base class BotAdapter) implement their own Adapter or pick the one we publish. For ASP.Net Apps (Core or Legacy) use the BotFrameworkAdapter and for tests use TestAdapter. Want to develop a windows service that runs in-box use a ServiceAdapter that read stuff from shared memory etc. Till the point you can pass IActivity to the BotAdapter class it doesn't care. No need to put the burden of understanding the ClaimsPrincipal on Middlewares or Adapters. Till the point you can call RunPipeline from your Adapter, Core package doesn't care.

After reading the DCR and doing some thinking of my own I came up with a plan of how I think we can split the dlls up and how people can use them in different scenarios.

**How does the split of dlls work**
1) Adapter.X -> BotAdapterFramework and BotFrameworkBotContext move to one package (or two one for ASP.Net, one for ASPNetCore). Similarily ConsoleAdapter into one and TestAdapter into another.
2) Connector.Http -> Contains contents of current Bot.Connector this could be merged into the BotFrameworkAdapter dll as well (though maybe not). This package allows you to connect to internet based services while not imposing you to use ASP.NET (want to write a bot which processes message from a queue and reply to threads this is what you will need, and write your own Adapter).
3) Bot.Builder.Core -> BotAdapter + IMiddleware + IStorage + Bot.Schema (a few more stuff). This is completely platform agnostic and doesn't even assume that the pipeline is HTTP.

**How do these pieces work in different scenarios**
- ASPNetCore -> You install BotFrameworkAdapter.ASPNetCore which in turns install 2) and 3) and In controller you get a BotAdapter instance and call ProcessActivity on the adapter.
- UnitTest -> You install TestAdapter and pass that instance into where you would normally pass other adapters and it should just work.
- Console\WPF\WinFormApp -> Use an appropriate Adapter (dlladapter?) which in turn uses 3) and in SendActivities function of Adapter you can serialize it to string on console, or show it in UI etc.
- Windows Server -> This follows a path similar the Console app with difference in how one would get input.

Let me know the issues you see with this approach.

**Comment by drub0y** [comment]
I think we're headed in the same direction! 👍

### RE: IBot
The DCR that intros `IBot` (#212) what I see as the first step we need to take in this direction. The reason for `IBot` is that I think it's important to differentiate between a piece of middleware, which has pipeline semantics, and the thing at "end of the pipe" which is what `IBot` represents. I think of it as analogous to writing a controller in ASP.NET. It removes some the mental "tax" of middleware from the average user who just wants to sit down and write a bot that exhibits their desired behavior (really helps deliver into the "pit of success"). 

Next, like the middleware, it gives a consistent API for bot development no matter what the host is. People can write their `IBot` and host it in ASP.NET 4.x today, but then when they're ready to make the shift to ASP.NET Core they don't have to go rewrite anything other than the hosting configuration and their bot "just works". Want to host it in a Windows service? Fine, just reference the same `IBot` implementation and configure the Windows service to expose it. Azure Functions is a whole other beast and the story for that should be absolutely be considered as well.

Finally, it actually takes the first step in enabling dependency injection/lifetime semantics which .NET developers are used to and have been missing up to this point. 

### Buffer, then rework
If we can at least deliver that abstraction now, the average developer experience can start there and it provides a buffer to go rework the upstream stuff. If we can dial in the host configuration APIs in that DCR it won't matter how much the implementation of that core architecture changes. Even if how hosting itself works changes, _the host specific configuration APIs_ shouldn't drift too far.

For example, @cmayomsft can go tweak his Promptly library so that it "starts" from the `IBot` abstraction, start to use the new host configuration APIs and delete all his own self hosting logic that is tightly coupled to the adapter/middleware design right now. He will not have to worry about reworking his entire library on the next big change because the only _real_ binding he now has in his library is a method signature of `OnReceiveAsync(BotContext)` which seems like it's probably pretty locked in no matter what we're doing upstream from this point fwd. 😄

### Next steps?
IMHO the next step is continuing this discussion "live" somewhere with a white boarding session to get the basic architecture laid out and deliver a simple architecture diagram that we can all scrutinize and agree on before moving the code in any specific direction. It feels pre-mature to me to get caught up in the packaging concerns before the target architecture is well understood. There are many options to "package things up" in the end including using a meta-package that could pull in many smaller packages (just as ASP.NET Core does) to, once again, provide that "pit of success" experience. If the meta-package approach leaves a bad taste in people's mouths for some reason, that's fine... we can still choose to have many projects and deliver those projects' multiple assemblies via a single NuGet package.

**Comment by cleemullins** [comment]
I think, based on the PR that was merged on Friday, that we're there. 

@RamjotSingh - I get your concerns around over-engineering, but the resulting code seems clean, easy to manipulate, and clear to understand. I don't see that we've lost anything with regards to flexiblity, feel we've gained the ability to have a Bot work with multiple adapters, and generally offered a much cleaner integration with ASP.Net WebAPI and ASP.Net Core.

**Comment by cleemullins** [comment]
For now, the design we have looks like what's going to take us through //Build. IT's been pretty thoroughly tested by folks at hacks, and people seem comfortable with it.
